design/secure-credentials-store.md

```diff
@@ -21,51 +21,120 @@ Credentials include:
+     * @param vaultSecretId Identifier of the secret. It corresponds to the value entered by user
+     *        in the respective configuration, which in turn is obtained from the vault when storing
+     *        the secret.
+     * @param purpose Purpose of the secret, set by keycloak according to the caller context.
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(KeycloakContext context, String vaultSecretId, String purpose);
+}
+```
+
+Vault configuration has to contain mapping between purpose and the allowed provider IDs sorted by priority to be used for this purpose, e.g. `realm-key => HSM, vault-1`, `smtp => vault-1`, `default => vault-2, HSM` would mean that:
+* realm keys should primarily be obtained from a provider with id `HSM` and then from a provider with ID `vault-1` (nothing else)
+* For SMTP secrets only `vault-1` should be consulted, and
+* For any other purpose, `vault-2` and `HSM` should be used to retrieve secrets in that order.
+
+### Vault Service Provider
+
+Individual vault services have to provide the secrets based on the context (e.g. realm) and the ID.
+They target the individual implementations.
+
+In Java pseudocode:
+
+```java
+public interface VaultServiceProvider {
+    /**
+     * Retrieves a secret from vault. The implementation should respect at least the realm ID
+     * to separate the secrets within the vault. If the secret is retrieved successfully,
+     * it is returned
+     *
+     * @param context Context that contains info about current request, to provide e.g. realm separation
+     * @param vaultSecretId Identifier of the secret. It corresponds to the value entered by user
+     *        in the respective configuration, which in turn is obtained from the vault when storing
+     *        the secret.
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(KeycloakContext context, String vaultSecretId);
+}
+```
+
+#### Default vault service providers to consider
+
+This is the list of vaults considered for a default implementations (TODO: to be shortlisted):
+
+* [Kubernetes](https://kubernetes.io/docs/concepts/configuration/secret/) / [OpenShift](https://docs.openshift.com/container-platform/4.1/nodes/pods/nodes-pods-secrets.html) secrets
+* Plain text (for testing purposes)
+* Database - use application level encryption to securely store secrets in the database
+  in a portable way. This would require dealing with keys as well as key rotation.
+* [HashiCorp vault](https://www.vaultproject.io/)
+* [Keywhiz](https://square.github.io/keywhiz/)
+* [Elytron Credential Store](https://docs.wildfly.org/17/WildFly_Elytron_Security.html#secure-credential-storage)
```

Comment:
What does Elytron actually support? Is it just a basic vault, or does it delegate to others? If just a basic vault it should be dropped from being considered, but if it enables support for multiple vault providers itself, than it could make a lot of sense.

Reply:
As far as I know, this is just a standalone vault that uses files. It doesn't allow us to delegate to other vaults.